File: src/main/java/com/tibiawiki/domain/objects/HuntingPlaceSkills.java
package com.tibiawiki.domain.objects;

import com.fasterxml.jackson.annotation.JsonBackReference;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.List;

@Getter
@Builder
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Component
public class HuntingPlaceSkills {

    private final String areaname;
    private final String lvlknights;
    private final String lvlpaladins;
    private final String lvlmages;
    private final String skknights;
    private final String skpaladins;
    private final String skmages;
    private final String defknights;
    private final String defpaladins;
    private final String defmages;

    private HuntingPlaceSkills() {
        this.areaname = null;
        this.lvlknights = null;
        this.lvlpaladins = null;
        this.lvlmages = null;
        this.skknights = null;
        this.skpaladins = null;
        this.skmages = null;
        this.defknights = null;
        this.defpaladins = null;
        this.defmages = null;
        this.huntingPlace = null;
    }

    @JsonBackReference
    private HuntingPlace huntingPlace;

    public static List<String> fieldOrder() {
        return Arrays.asList("areaname", "lvlknights", "lvlpaladins", "lvlmages", "skknights", "skpaladins", "skmages",
                "defknights", "defpaladins", "defmages");
    }
}

File: src/main/kotlin/com/tibiawiki/domain/objects/HuntingPlaceSkills.kt
package com.tibiawiki.domain.objects

import com.fasterxml.jackson.annotation.JsonBackReference

/**
 * Per-area Hunt skill table. Same knight / paladin / mage columns as [HuntingPlace];
 * no Monk fields until the wiki Infobox Hunt Skills template adds them.
 */
data class HuntingPlaceSkills(
    val areaname: String? = null,
    val lvlknights: String? = null,
    val lvlpaladins: String? = null,
    val lvlmages: String? = null,
    val skknights: String? = null,
    val skpaladins: String? = null,
    val skmages: String? = null,
    val defknights: String? = null,
    val defpaladins: String? = null,
    val defmages: String? = null,
    @JsonBackReference
    val huntingPlace: HuntingPlace? = null
) {
    companion object {
        fun fieldOrder(): List<String> {
            return listOf(
                "areaname",
                "lvlknights",
                "lvlpaladins",
                "lvlmages",
                "skknights",
                "skpaladins",
                "skmages",
                "defknights",
                "defpaladins",
                "defmages"
            )
        }
    }
}
